From 00c8b2a9cd94d0a2c71ccf4985b1789526310fce Mon Sep 17 00:00:00 2001
From: Brian Delwiche <delwiche@google.com>
Date: Tue, 6 May 2025 17:51:32 +0000
Subject: [PATCH] Check for existing discovery database in
 bta_hf_client_cb_init

Researcher finds an RCE related to HF client discovery database.

Changes against the related issue b/291281168 (guarded by flag below)
partially resolve this issue, but it is necessary to add a check in
another location.

Bug: 374746961
Bug: 356201480
Test: m libbluetooth
Test: researcher POC
Ignore-AOSP-First: security
Tag: #security
Merged-In: Ie6329743c8acdc72154b9e3774ca53e840c85478
Merged-In: I085e65db0d6f8c7c8c474e193ce7d676d0a9b398
Merged-In: I69b2e80f0dd8682dcc8862b4b9aa7845c48c3bb8
Change-Id: I69b2e80f0dd8682dcc8862b4b9aa7845c48c3bb8
---
 system/bta/Android.bp                      | 8 +++++---
 system/bta/hf_client/bta_hf_client_main.cc | 8 ++++++++
 system/bta/test/bta_dip_test.cc            | 1 -
 3 files changed, 13 insertions(+), 4 deletions(-)

diff --git a/system/bta/Android.bp b/system/bta/Android.bp
index 939f0bec72..85e2389a7f 100644
--- a/system/bta/Android.bp
+++ b/system/bta/Android.bp
@@ -161,6 +161,7 @@ cc_test {
     test_suites: ["device-tests"],
     srcs: [
         ":TestMockStackBtm",
+        ":TestMockStackSdp",
         ":TestCommonMockFunctions",
         "test/bta_hf_client_test.cc",
         "test/bta_dm_cust_uuid_test.cc",
@@ -204,9 +205,9 @@ cc_test {
     test_suites: ["device-tests"],
     srcs: [
         ":TestCommonMockFunctions",
-	":TestMockDevice",
-	":TestMockStack",
-	":TestMockBtif",
+        ":TestMockDevice",
+        ":TestMockStack",
+        ":TestMockBtif",
         "test/bta_hf_client_security_test.cc",
     ],
     shared_libs: [
@@ -227,6 +228,7 @@ cc_test {
         "libbt-common",
     ],
 }
+
 cc_test {
     name: "bt_host_test_bta",
     defaults: [
diff --git a/system/bta/hf_client/bta_hf_client_main.cc b/system/bta/hf_client/bta_hf_client_main.cc
index 441e3247f4..2533cd39c5 100644
--- a/system/bta/hf_client/bta_hf_client_main.cc
+++ b/system/bta/hf_client/bta_hf_client_main.cc
@@ -305,6 +305,14 @@ void bta_hf_client_cb_init(tBTA_HF_CLIENT_CB* client_cb, uint16_t handle) {
   client_cb->enabled_hf_indicators.clear();
   client_cb->peer_hf_indicators.clear();
 
+  if (client_cb->p_disc_db) {
+    if (!SDP_CancelServiceSearch(client_cb->p_disc_db)) {
+      APPL_TRACE_WARNING("Unable to cancel SDP service discovery peer: %s",
+                         client_cb->peer_addr.ToString().c_str());
+    }
+    osi_free_and_reset((void**)&client_cb->p_disc_db);
+  }
+
   // Memset the rest of the block
   // memset(client_cb, 0, sizeof(tBTA_HF_CLIENT_CB));
   *client_cb = {};
diff --git a/system/bta/test/bta_dip_test.cc b/system/bta/test/bta_dip_test.cc
index 3e71e2e6e1..d1f985e9e3 100644
--- a/system/bta/test/bta_dip_test.cc
+++ b/system/bta/test/bta_dip_test.cc
@@ -20,7 +20,6 @@
 
 #include "bta/sdp/bta_sdp_act.cc"
 #include "main/shim/metrics_api.h"
-#include "stack/sdp/sdp_api.cc"
 #include "types/bluetooth/uuid.h"
 #include "types/raw_address.h"
 
-- 
2.51.0.362.g3baa7cb742-goog

